File: src/io_util/pipes_util/create_pipes.go
package pipes_util

import (
	"errors"
	"fmt"
	"os"
	"syscall"
)

func handle(e error) {
	if e != nil {
		panic(e)
	}
}

func CreatePipe(path string, mode string) *os.File {
	// Golang opens it in non-blocking mode (to prevent deadlock on opening)
	var flags int
	if mode == "send" {
		// The writer should be in blocking mode so that it waits for the reader to open the pipe
		flags = syscall.O_CREAT | syscall.O_WRONLY
	}
	if mode == "recv" {
		// The reader needs to open the pipe and be in non-blocking mode
		err := syscall.Mkfifo(path, 0666)
		if err != nil && !errors.Is(err, os.ErrExist) {
			panic(err)
		}
		flags = syscall.O_NONBLOCK | syscall.O_CREAT | syscall.O_RDONLY
	}

	fmt.Println("Trying to open pipe")
	pipe, err := os.OpenFile(path, flags, os.ModeNamedPipe)
	for os.IsNotExist(err) {
		pipe, err = os.OpenFile(path, flags, os.ModeNamedPipe)
	}
	handle(err)

	fmt.Println("Opened pipe")
	return pipe
}

File: src/io_util/pipes_util/create_pipes.py
import os
import time


def create_pipe(path: str, mode: str) -> int:
    options = 0
    # Open pipes as blocking on the python side
    if mode == "send":
        # The writer should be in blocking mode so that it waits for the reader to open the pipe
        options = os.O_WRONLY
    elif mode == "recv":
        # The reader needs to open the pipe and be in non-blocking mode
        try:
            os.mkfifo(path, 0o666)
        except FileExistsError:
            pass
        options = os.O_NONBLOCK | os.O_RDONLY

    print("os.open() pipe")
    pipe = 0
    err = FileNotFoundError()
    while type(err) is FileNotFoundError:
        try:
            pipe = os.open(path, options)
            # break out of error loop
            err = 0
        except FileNotFoundError as e:
            print("Pipe wasn't created for", mode)
            err = e
            time.sleep(0.5)
    print("Pipe created")
    return pipe
